types/instance.d.ts

```diff
@@ -499,6 +499,18 @@ export interface FastifyInstance<
     hook: onReadyAsyncHookHandler,
   ): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
 
+  /**
+  * Triggered when fastify.listen() is invoked to start the server. It is useful when plugins need a "onListen" event, for example to run logics after the server start listening for requests.
+  */
+  addHook(
+    name: 'onListen',
+    hook: onListenHookHandler
+  ): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
+
+  addHook(
+    name: 'onListen',
+    hook: onListenAsyncHookHandler,
+  ): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
```

Comment:
Infer the hook handler for onListen correcly, compare with #4945 

```suggestion
  addHook<
    Fn extends onListenHookHandler | onListenAsyncHookHandler = onListenHookHandler
  >(
    name: 'onListen',
    hook: Fn extends unknown ? Fn extends (...args: any) => Promise<any> ? onListenAsyncHookHandler : onListenHookHandler : Fn,
  ): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
```

Reply:
Should we wait till #4945 is merged before implementing this change?

Reply:
if #4945 is merged before this implementation, than adapt, if not let it  be, and #4945 has to merge main and adapt